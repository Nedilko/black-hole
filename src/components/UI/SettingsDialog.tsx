import { useState } from 'react';
import { ISettings } from '../../App';
import NumberInput from './NumberInput';

type PropsType = {
  onStartGame: (settings: ISettings) => void;
  settings: ISettings;
};

const SettingsDialog = ({ onStartGame, settings }: PropsType) => {
  const handleApply = () => {
    console.log('apply');
    onStartGame({
      size: {
        width,
        height,
      },
      holesCount,
    });
  };

  const [width, setWidth] = useState<number>(settings.size.width);
  const [height, setHeight] = useState<number>(settings.size.height);
  const [holesCount, setHolesCount] = useState<number>(settings.holesCount);

  return (
    <div className="flex flex-col p-4">
      <div className="flex flex-col justify-center items-center">
        <h2 className="text-4xl text-gray-300/80 uppercase mb-4">Settings</h2>
        <div className="flex flex-row mt-4 gap-x-6">
          <NumberInput
            label="Width:"
            value={width}
            min={2}
            max={20}
            handleChange={setWidth}
          />
          <NumberInput
            label="Height:"
            min={2}
            max={20}
            value={height}
            handleChange={setHeight}
          />
        </div>
        <div className="flex mt-4">
          <NumberInput
            label="Holes count:"
            min={1}
            max={width * height - 2}
            value={holesCount}
            handleChange={setHolesCount}
          />
        </div>
      </div>
      <div className="flex justify-center mt-12">
        <button
          className="pb-1 text-5xl text-gray-300/80 uppercase hover:text-shadow hover:text-gray-300 transition-all duration-150"
          onClick={handleApply}
        >
          Start
        </button>
      </div>
    </div>
  );
};

export default SettingsDialog;
