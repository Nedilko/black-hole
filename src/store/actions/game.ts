import { createSlice } from '@reduxjs/toolkit';
import type { FieldSize } from './field';
import type { AppDispatch } from '..';
import { fieldActions } from './field';

export interface GameState {
  showGameField: boolean;
  showGameControls: boolean;
  remainingCellsCount: number;
  totalCellsCount: number;
  showTime: boolean;
  showCounter: boolean;
  showSettings: boolean;
  time: string;
}

const defaultStateValue: GameState = {
  showGameField: false,
  showGameControls: false,
  remainingCellsCount: 0,
  totalCellsCount: 0,
  showTime: false,
  showCounter: false,
  showSettings: true,
  time: '00:00',
};

const gameSlice = createSlice({
  name: 'game',
  initialState: defaultStateValue,
  reducers: {
    showGameField: (state) => {
      state.showGameField = true;
    },
    hideGameField: (state) => {
      state.showGameField = false;
    },
    showGameControls: (state) => {
      state.showGameControls = true;
    },
    hideGameControls: (state) => {
      state.showGameControls = false;
    },
    showCounter: (state) => {
      state.showCounter = true;
    },
    hideCounter: (state) => {
      state.showCounter = false;
    },
    showTime: (state) => {
      state.showTime = true;
    },
    hideTime: (state) => {
      state.showTime = false;
    },
    showSettings: (state) => {
      state.showSettings = true;
    },
    hideSettings: (state) => {
      state.showSettings = false;
    },
    finishGame: (state) => {
      state.showGameControls = true;
      state.showTime = false;
      state.showCounter = false;
    },
  },
});

export const showMainMenu = (): any => (dispatch: AppDispatch) => {
  dispatch(gameSlice.actions.hideGameField());
  dispatch(gameSlice.actions.hideGameControls());
  dispatch(gameSlice.actions.hideCounter());
  dispatch(gameSlice.actions.hideTime());
  dispatch(gameSlice.actions.showSettings());
};

export const restartGame = (): any => (dispatch: AppDispatch) => {
  dispatch(fieldActions.renewField());
  dispatch(gameSlice.actions.showCounter());
  dispatch(gameSlice.actions.showTime());
  dispatch(gameSlice.actions.hideGameControls());
};

export const startGame =
  (size: FieldSize, holesCount: number): any =>
  (dispatch: AppDispatch): any => {
    dispatch(fieldActions.setupField({ size, holesCount }));
    dispatch(gameSlice.actions.showGameField());
    dispatch(gameSlice.actions.hideGameControls());
    dispatch(gameSlice.actions.showCounter());
    dispatch(gameSlice.actions.showTime());
    dispatch(gameSlice.actions.hideSettings());
  };

export default gameSlice.reducer;

export const gameActions = gameSlice.actions;
