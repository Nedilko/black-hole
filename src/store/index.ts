import {
  configureStore,
  combineReducers,
  PreloadedState,
} from '@reduxjs/toolkit';
import gameReducer from './gameSlice';
import fieldReducer from './fieldSlice';

const rootReducer = combineReducers({
  game: gameReducer,
  field: fieldReducer,
});

export function setupStore(preloadedState?: PreloadedState<RootState>) {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
  });
}
export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];
