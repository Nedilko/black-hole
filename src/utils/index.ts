import { getTime } from './time';
import {
  readFromLocalstorage,
  writeToLocalstorage,
} from './localstorageAdapter';
import { getDefaultResults } from './game';
import { loadResults, writeResults } from './dataAdapter';

export {
  getTime,
  readFromLocalstorage,
  writeToLocalstorage,
  getDefaultResults,
  loadResults,
  writeResults,
};
