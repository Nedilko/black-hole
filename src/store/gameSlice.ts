import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import type { FieldSize } from './fieldSlice';
import type { AppDispatch } from '.';
import { renewField, setupField } from './fieldSlice';
import { loadResults } from '../utils';

export enum DIFFUCULTY {
  EASY,
  MEDIUM,
  HARD,
  CUSTOM,
}

export interface Results {
  [DIFFUCULTY.EASY]: number;
  [DIFFUCULTY.MEDIUM]: number;
  [DIFFUCULTY.HARD]: number;
  [DIFFUCULTY.CUSTOM]: number;
}

export interface GameState {
  showGameField: boolean;
  showGameControls: boolean;
  remainingCellsCount: number;
  totalCellsCount: number;
  showTime: boolean;
  showCounter: boolean;
  showSettings: boolean;
  time: number;
  isStarted: boolean;
  isFinished: boolean;
  isWon: boolean;
  difficulty: DIFFUCULTY;
  results: Results;
}

const defaultStateValue: GameState = {
  showGameField: false,
  showGameControls: false,
  remainingCellsCount: 0,
  totalCellsCount: 0,
  showTime: false,
  showCounter: false,
  showSettings: true,
  time: 0,
  isStarted: false,
  isFinished: false,
  isWon: false,
  difficulty: DIFFUCULTY.EASY,
  results: loadResults(),
};

const gameSlice = createSlice({
  name: 'game',
  initialState: defaultStateValue,
  reducers: {
    finishGame(state) {
      state.showGameControls = true;
      state.isFinished = true;
      state.isStarted = false;
    },
    showMainMenu(state) {
      state.showGameField = false;
      state.showGameControls = false;
      state.showCounter = false;
      state.showTime = false;
      state.showSettings = true;
      state.time = 0;
      state.isFinished = false;
      state.isWon = false;
    },
    showGameField(state) {
      state.showGameControls = false;
      state.showCounter = true;
      state.showTime = true;
      state.showSettings = false;
      state.showGameField = true;
      state.time = 0;
      state.isFinished = false;
      state.isWon = false;
    },
    updateTime(state) {
      state.time++;
    },
    startCountdown(state) {
      state.isStarted = true;
    },
    winGame(state) {
      state.isWon = true;
      state.showGameField = false;
      state.showGameControls = true;
      state.isFinished = true;
      state.isStarted = false;
      if (
        state.results[state.difficulty] > state.time ||
        state.results[state.difficulty] === 0
      ) {
        state.results[state.difficulty] = state.time;
      }
    },
    registerDifficulty(state, action: PayloadAction<DIFFUCULTY>) {
      state.difficulty = action.payload;
    },
  },
});

export default gameSlice.reducer;

export const {
  finishGame,
  showMainMenu,
  updateTime,
  winGame,
  registerDifficulty,
  startCountdown,
} = gameSlice.actions;

export const restartGame = () => (dispatch: AppDispatch) => {
  dispatch(renewField());
  dispatch(gameSlice.actions.showGameField());
};

export const startGame =
  (size: FieldSize, holesCount: number) => (dispatch: AppDispatch) => {
    dispatch(setupField({ size, holesCount }));
    dispatch(gameSlice.actions.showGameField());
  };
