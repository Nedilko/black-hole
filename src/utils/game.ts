import { DIFFUCULTY, Results } from '../store/gameSlice';

export const getDefaultResults = (): Results => {
  return {
    [DIFFUCULTY.EASY]: 0,
    [DIFFUCULTY.MEDIUM]: 0,
    [DIFFUCULTY.HARD]: 0,
    [DIFFUCULTY.CUSTOM]: 0,
  };
};
