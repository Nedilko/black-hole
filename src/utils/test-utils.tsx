import { RenderOptions } from '@testing-library/react';
import { PropsWithChildren } from 'react';
import { Provider } from 'react-redux';
import { PreloadedState } from 'redux';
import { AppStore, RootState, setupStore } from '../store';
import { render } from '@testing-library/react';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  initialState?: PreloadedState<RootState>;
  store?: AppStore;
}

export const renderWithProviders = (
  ui: React.ReactElement,
  {
    initialState = {},
    store = setupStore(initialState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }
  return {
    store,
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  };
};
