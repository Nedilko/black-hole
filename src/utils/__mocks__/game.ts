import type { Results } from '../../store/gameSlice';
import { DIFFUCULTY } from '../../utils/game';

export const getDefaultResults = (): Results => {
  return {
    [DIFFUCULTY.EASY]: 0,
    [DIFFUCULTY.MEDIUM]: 1,
    [DIFFUCULTY.HARD]: 2,
    [DIFFUCULTY.CUSTOM]: 3,
  };
};
