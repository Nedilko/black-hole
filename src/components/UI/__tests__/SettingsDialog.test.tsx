import userEvent from '@testing-library/user-event';
import { renderWithProviders } from '../../../utils/test-utils';
import SettingsDialog from '../SettingsDialog';
import { fireEvent, screen } from '@testing-library/react';
import * as actions from '../../../store/gameSlice';
import * as hooks from '../../../hooks';
import { DIFFUCULTY } from '../../../utils/game';

jest.mock('../../../hooks', () => ({
  ...jest.requireActual('../../../hooks'),
  useAppDispatch: jest.fn(),
}));
jest.mock('../../../hooks');

const registerDifficultyAction = jest.spyOn(actions, 'registerDifficulty');
const startGameAction = jest.spyOn(actions, 'startGame');

describe('SettingsDialog', () => {
  afterEach(() => {
    jest.clearAllMocks();
    registerDifficultyAction.mockReset();
    startGameAction.mockReset();
  });

  it('should render', () => {
    const { container } = renderWithProviders(<SettingsDialog />);
    expect(container).toMatchSnapshot();
  });

  it('should handle start button click', () => {
    jest.spyOn(hooks, 'useAppDispatch').mockReturnValue(jest.fn());
    renderWithProviders(<SettingsDialog />);

    userEvent.click(screen.getByRole('button'));

    expect(startGameAction).toHaveBeenCalledTimes(1);
    expect(registerDifficultyAction).toHaveBeenCalledTimes(1);
  });

  it('should handle difficulty change', () => {
    jest.spyOn(hooks, 'useAppDispatch').mockReturnValue(jest.fn());
    renderWithProviders(<SettingsDialog />);

    userEvent.selectOptions(screen.getByRole('combobox'), 'Medium');
    userEvent.click(screen.getByRole('button'));

    expect(registerDifficultyAction).toHaveBeenCalledWith(DIFFUCULTY.MEDIUM);
  });

  it('shouldn handle inputs change if diffuculty selected', () => {
    jest.spyOn(hooks, 'useAppDispatch').mockReturnValue(jest.fn());
    renderWithProviders(<SettingsDialog />);

    userEvent.selectOptions(screen.getByRole('combobox'), 'Medium');
    userEvent.click(screen.getByRole('button'));

    expect(startGameAction).toHaveBeenCalledWith({ width: 15, height: 15 }, 20);

    userEvent.selectOptions(screen.getByRole('combobox'), 'Hard');
    userEvent.click(screen.getByRole('button'));

    expect(startGameAction).toHaveBeenCalledWith({ width: 20, height: 20 }, 30);

    userEvent.selectOptions(screen.getByRole('combobox'), 'Hard');
    userEvent.selectOptions(screen.getByRole('combobox'), 'Easy');
    userEvent.click(screen.getByRole('button'));

    expect(startGameAction).toHaveBeenCalledWith({ width: 8, height: 8 }, 10);
  });

  it("shouldn't handle inputs change if diffuculty set to custom", () => {
    jest.spyOn(hooks, 'useAppDispatch').mockReturnValue(jest.fn());
    renderWithProviders(<SettingsDialog />);

    userEvent.selectOptions(screen.getByRole('combobox'), 'Custom');
    userEvent.click(screen.getByRole('button'));

    expect(startGameAction).toHaveBeenCalledWith({ width: 8, height: 8 }, 10);
  });

  it('should handle width, height, holes count change', () => {
    jest.spyOn(hooks, 'useAppDispatch').mockReturnValue(jest.fn());
    renderWithProviders(<SettingsDialog />);

    fireEvent.change(screen.getByLabelText(/width/i), {
      target: { value: '12' },
    });
    fireEvent.change(screen.getByLabelText(/height/i), {
      target: { value: '13' },
    });
    fireEvent.change(screen.getByLabelText(/holes count/i), {
      target: { value: '14' },
    });
    userEvent.click(screen.getByRole('button'));

    expect(startGameAction).toHaveBeenCalledWith({ width: 12, height: 13 }, 14);
  });

  it("shouldn't change width if width and height less than 5", () => {
    jest.spyOn(hooks, 'useAppDispatch').mockReturnValue(jest.fn());

    renderWithProviders(<SettingsDialog />);

    fireEvent.change(screen.getByLabelText(/width/i), {
      target: { value: '4' },
    });
    fireEvent.change(screen.getByLabelText(/height/i), {
      target: { value: '4' },
    });
    userEvent.click(screen.getByRole('button'));

    expect(startGameAction).toHaveBeenCalledWith({ width: 8, height: 8 }, 10);
  });

  it("shouldn't change width if width and height greater than 20", () => {
    jest.spyOn(hooks, 'useAppDispatch').mockReturnValue(jest.fn());

    renderWithProviders(<SettingsDialog />);

    fireEvent.change(screen.getByLabelText(/width/i), {
      target: { value: '21' },
    });
    fireEvent.change(screen.getByLabelText(/height/i), {
      target: { value: '21' },
    });
    userEvent.click(screen.getByRole('button'));

    expect(startGameAction).toHaveBeenCalledWith({ width: 8, height: 8 }, 10);
  });

  it("shouldn't change holes count if it greater than width * height - 1", () => {
    jest.spyOn(hooks, 'useAppDispatch').mockReturnValue(jest.fn());
    const width = 8;
    const height = 8;

    renderWithProviders(<SettingsDialog />);

    fireEvent.change(screen.getByLabelText(/holes count/i), {
      target: { value: width * height - 1 },
    });
    userEvent.click(screen.getByRole('button'));

    expect(startGameAction).toHaveBeenCalledWith({ width, height }, 10);
  });

  it("shouldn't change holes count if it less than 1", () => {
    jest.spyOn(hooks, 'useAppDispatch').mockReturnValue(jest.fn());
    const width = 8;
    const height = 8;

    renderWithProviders(<SettingsDialog />);

    fireEvent.change(screen.getByLabelText(/holes count/i), {
      target: { value: 0 },
    });
    userEvent.click(screen.getByRole('button'));

    expect(startGameAction).toHaveBeenCalledWith({ width, height }, 10);
  });
});
