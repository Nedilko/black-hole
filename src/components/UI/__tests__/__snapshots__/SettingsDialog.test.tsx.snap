// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsDialog should render 1`] = `
<div>
  <div
    class="flex flex-col p-4 w-full max-w-[400px]"
  >
    <div
      class="flex flex-col justify-center items-center"
    >
      <div
        class="flex flex-row mt-4 gap-x-6 w-full"
      >
        <label
          class="flex text-gray-300 uppercase relative w-full"
        >
          <svg
            class="absolute translate-y-1/2 right-0 cursor-pointer text-3xl pointer-events-none"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
            <path
              d="M7.41 8.59L12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
            />
          </svg>
          <div
            class="text-sm absolute top-1 left-2 text-gray-400"
          >
            difficulty
          </div>
          <select
            class="appearance-none outline outline-1 hover:outline-2 outline-slate-700 focus:outline-gray-400 outline-offset-2 w-full text-xl font-normal bg-slate-900 px-2 pt-6 pb-1 rounded transition-all duration-150 hover:cursor-pointer"
            data-testid="presets-dropdown"
            name="presets-dropdown"
          >
            <option
              value="0"
            >
              Easy
            </option>
            <option
              value="1"
            >
              Medium
            </option>
            <option
              value="2"
            >
              Hard
            </option>
            <option
              value="3"
            >
              Custom
            </option>
          </select>
        </label>
      </div>
      <div
        class="flex flex-row mt-4 gap-x-6 w-full"
      >
        <label
          class="flex text-gray-300 uppercase relative w-full"
        >
          <svg
            class="absolute top-0 right-0 cursor-pointer text-3xl hover:scale-110 transition-all duration-150"
            data-testid="increment"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M7.41 15.41L12 10.83l4.59 4.58L18 14l-6-6-6 6z"
            />
          </svg>
          <svg
            class="absolute bottom-0 right-0 cursor-pointer text-3xl hover:scale-110 transition-all duration-150"
            data-testid="decrement"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
            <path
              d="M7.41 8.59L12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
            />
          </svg>
          <div
            class="text-sm absolute top-1 left-2 text-gray-400 select-none"
          >
            Width:
          </div>
          <input
            class="appearance-none outline outline-1 hover:outline-2 outline-slate-700 focus:outline-gray-400 outline-offset-2 w-full text-xl font-semibold bg-slate-900 px-2 pt-6 pb-1 rounded transition-all duration-150"
            max="20"
            min="2"
            type="number"
            value="8"
          />
        </label>
        <label
          class="flex text-gray-300 uppercase relative w-full"
        >
          <svg
            class="absolute top-0 right-0 cursor-pointer text-3xl hover:scale-110 transition-all duration-150"
            data-testid="increment"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M7.41 15.41L12 10.83l4.59 4.58L18 14l-6-6-6 6z"
            />
          </svg>
          <svg
            class="absolute bottom-0 right-0 cursor-pointer text-3xl hover:scale-110 transition-all duration-150"
            data-testid="decrement"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
            <path
              d="M7.41 8.59L12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
            />
          </svg>
          <div
            class="text-sm absolute top-1 left-2 text-gray-400 select-none"
          >
            Height:
          </div>
          <input
            class="appearance-none outline outline-1 hover:outline-2 outline-slate-700 focus:outline-gray-400 outline-offset-2 w-full text-xl font-semibold bg-slate-900 px-2 pt-6 pb-1 rounded transition-all duration-150"
            max="20"
            min="2"
            type="number"
            value="8"
          />
        </label>
      </div>
      <div
        class="flex mt-4 w-full"
      >
        <label
          class="flex text-gray-300 uppercase relative w-full"
        >
          <svg
            class="absolute top-0 right-0 cursor-pointer text-3xl hover:scale-110 transition-all duration-150"
            data-testid="increment"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M7.41 15.41L12 10.83l4.59 4.58L18 14l-6-6-6 6z"
            />
          </svg>
          <svg
            class="absolute bottom-0 right-0 cursor-pointer text-3xl hover:scale-110 transition-all duration-150"
            data-testid="decrement"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
            <path
              d="M7.41 8.59L12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
            />
          </svg>
          <div
            class="text-sm absolute top-1 left-2 text-gray-400 select-none"
          >
            Holes count:
          </div>
          <input
            class="appearance-none outline outline-1 hover:outline-2 outline-slate-700 focus:outline-gray-400 outline-offset-2 w-full text-xl font-semibold bg-slate-900 px-2 pt-6 pb-1 rounded transition-all duration-150"
            max="62"
            min="1"
            type="number"
            value="10"
          />
        </label>
      </div>
    </div>
    <div
      class="flex justify-center mt-12"
    >
      <button
        class="pb-1 text-5xl text-gray-300/80 uppercase hover:text-shadow hover:text-gray-300 transition-all duration-150"
      >
        Start
      </button>
    </div>
  </div>
</div>
`;
